File: merge_sort.py
import math
import random

def merge_sort(to_sort):

    # base case to exit recursive loop
    if len(to_sort) == 1:
        return to_sort
    
    # the point to divide the list in half (ceil to accompany for odd length lists)
    half_idx = math.ceil(len(to_sort) / 2)

    l_half = to_sort[:half_idx]
    r_half = to_sort[half_idx:]

    sorted_l_half = merge_sort(l_half)
    sorted_r_half = merge_sort(r_half)
    return merge(sorted_l_half, sorted_r_half)

def merge(l_arr, r_arr):
    # print(f'Merging {l_arr} and {r_arr}', end=' ')

    merged = []

    while len(l_arr) > 0 or len(r_arr) > 0:

        if len(l_arr) == 0:
            merged.append(r_arr.pop(0))
            continue
        elif len(r_arr) == 0:
            merged.append(l_arr.pop(0))
            continue

        # compare and add the smaller value to merged
        if l_arr[0] <= r_arr[0]:
            merged.append(l_arr.pop(0))
        else:
            merged.append(r_arr.pop(0))

    # print(f'resulted in {merged}')
    return merged 

# generate list with 100 random integers and sort it
to_sort = [random.randint(0, 100) for x in range(100)] 
sorted_arr = merge_sort(to_sort)
print(to_sort, end='\n\n')
print(sorted_arr)

#answer = [1, 2, 4, 5, 7, 9, 18]
#print(f'New array {sorted_arr} vs Answer {answer}')
#print(sorted_arr == answer)

File: merge_sort.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <time.h>

void print_vector(std::vector<int> &to_print)
{
    for (std::size_t i=0; i < to_print.size(); i++)
    {
        std::cout << to_print[i] << " ";
    }
    std::cout << std::endl;
}

void merge(std::vector<int> &l_vect, std::vector<int> &r_vect, std::vector<int> &merged)
{
    std::size_t merged_idx = 0;
    std::size_t l_size = l_vect.size();
    std::size_t r_size = r_vect.size();
    std::size_t l_idx = 0;
    std::size_t r_idx = 0;

    for(merged_idx; merged_idx < merged.size(); merged_idx++)
    {
        if (l_idx == l_size) {
            merged[merged_idx] = r_vect[r_idx];
            r_idx++;
            continue;
        } else if (r_idx == r_size) {
            merged[merged_idx] = l_vect[l_idx];
            l_idx++;
            continue;
        }

        // compare and add the smaller value to merged
        if (l_vect[l_idx] <= r_vect[r_idx]) {
            merged[merged_idx] = l_vect[l_idx];
            l_idx++;
        } else {
            merged[merged_idx] = r_vect[r_idx];
            r_idx++;
        }
    }
}
void merge_sort(std::vector<int> &to_sort)
{
    if (to_sort.size() == 1)
        return;

    std::size_t half_idx = std::ceil((float)to_sort.size() / 2);

    // The iterator pointing to the half element of the vector
    std::vector<int>::iterator half_iter = to_sort.end() - (to_sort.size() - half_idx);

    // Iterators in C++ by The Cherno https://www.youtube.com/watch?v=SgcHcbQ0RCQ
    std::vector<int> l_half(to_sort.begin(), half_iter);
    std::vector<int> r_half(half_iter, to_sort.end());

    merge_sort(l_half);
    merge_sort(r_half);

    merge(l_half, r_half, to_sort);
}

int main()
{
    std::vector<int> to_sort;

    // Initialize random seed
    srand(time(NULL));

    for (int i = 0; i < 20; i++) {
        // Generate a random int with range 0~99
        int rand_int = rand() % 100;
        to_sort.push_back(rand_int);
    }

    merge_sort(to_sort);
    std::cout << "main() \n";
    print_vector(to_sort);
}